download/tasks.py

```diff
@@ -1,44 +1,151 @@
-import glob
-from os import path
+import fnmatch
+import hashlib
+import json
+import re
+import sys
 import tempfile
+import requests
+from datetime import datetime, timedelta
+from os import path, walk
 
 from invoke import task
 
 
+STATUS_FILE_NAME = "_files_status.json"
+
+
+def raw_files_status(ctx):
+    files_status_path = path.join(ctx.cache_dir, STATUS_FILE_NAME)
+
+    if path.isfile(files_status_path):
+        with open(files_status_path) as data:
+            return json.load(data)
+    else:
+        return dict()
+
+
+def get_file_status(ctx, filename):
+    res = raw_files_status(ctx).get(filename)
+
+    if res is None:
+        res = {"last_update": None, "md5": None}
+
+    if res["last_update"] is not None:
+        res["last_update"] = datetime.utcfromtimestamp(res["last_update"])
+
+    return res
+
+
+def save_file_status(ctx, filename, status):
+    files_status_path = path.join(ctx.cache_dir, STATUS_FILE_NAME)
+
+    if status is not None:
+        status["last_update"] = status["last_update"].timestamp()
+
+    files_status = raw_files_status(ctx)
+    files_status[filename] = status
+
+    with open(files_status_path, "w") as data:
+        json.dump(files_status, data)
+
+
+def needs_to_download(ctx, filename, max_age=None, md5_url=None):
+    """
+    Check if "filename" should be downloaded because it doesn't exist or is
+    older than "max_age" or its checksum has changed.
+    """
+    if not path.isfile(filename):
+        return True
+
+    file_status = get_file_status(ctx, filename)
+
+    if max_age is not None and (
+        file_status["last_update"] is None
+        or datetime.utcnow() - file_status["last_update"] > max_age
+    ):
+        return True
+
+    if md5_url is not None:
+        res = requests.get(md5_url)
```

Comment:
HTTP errors should be caught here.  
For example, the file should be downloaded if the md5 is not found.